packages/dds/tree/docs/wip/revertible-fork-api.md

```diff
@@ -0,0 +1,207 @@
+
+> The original "non-cloned" revertible, which came from the `"commitApplied"` event, is implicitly associated with the branch that raised the event.
+
+There is a particular scenario in which this is necessary for a user.
+If a user wants to create a new branch and allow that branch to undo past the commit at which it was created (i.e., undo "back into its parent branch"), then all the relevant revertibles from the parent branch must be cloned and the clones must be associated with the new branch. For example:
+
+1. Parent branch **P** appends commits _X_, _Y_, and _Z_.
+2. Child branch **C** branches off of the current head (_Z_) of **P**.
+3. The user wants to undo _Z_ and then _Y_, but on branch **C**, not on branch **P**.
+   This would leave **P** untouched and leave **C** with a state including only the commit _X_.
+
+In such a scenario the user must clone the revertibles that were acquired from **P** and associate them with **C**.
+Typically this would look something like the following.
+
+1. Parent branch **P** appends commits _X_, _Y_, and _Z_.
+    * The revertibles for _X_, _Y_ and _Z_ are appended to the undo stack for **P**.
+2. Child branch **C** branches off of the current head (_Z_) of **P**.
+    * The revertibles for _X_, _Y_ and _Y_ are cloned, and put at the beginning of the undo stack for **C**.
+3. The user wants to undo _Z_ and then _Y_, but on branch **C**, not on branch **P**.
+   This would leave **P** untouched and leave **C** with a state including only the commit _X_.
+    * The user pops and reverts the top two revertibles on **C**'s undo stack (and may or may not dispose them, depending on whether or not they might be needed again later).
+
+> When a branch is disposed, all of the revertibles that were cloned with respect to it are automatically disposed too.
+
+### Cloning
+
+Revertibles are cloned via a `clone()` method on the revertible object.
+
+```ts
+interface Revertible {
+	revert(): void;
+	dispose(): void;
+	clone(branch?: Branch): Revertible;
+}
+```
+
+Note the optional `branch` parameter to the clone method.
+If a branch is provided, then the resulting clone is associated with that branch (recall that every revertible is associated with one branch, which is the branch that the revert is applied to when `revert()` is called);
+If a branch is not provided, then the clone is associated with the same branch as the original revertible.
+The cloned revertibles may be disposed, just like the original revertibles.
+Disposing a clone doesn't affect original.
+
+There is one additional consequence of associating a revertible with a branch.
+When a branch is disposed, all revertibles that are associated with it are also automatically disposed.
+This makes sense because those revertibles could not possible do anything thereafter (they can't modify a disposed branch).
```

Comment:
This is not quite true because you could still associated them with a new branch by cloning them. I don't think that's a use case we ought to worry about, but just FYI.

Reply:
Interesting. If you did that, then the clones would be just as useless, no? They can't revert either, because the repair data is gone. That makes me think we should throw an error if you try to clone a disposed revertible. Technically we only need to throw an error if _all_ revertibles for that commit are disposed, but that's a confusing distinction.

Reply:
Oh, but that's exactly what you are pointing out - there can be other outstanding revertibles for that commit but for different branches. Yea, kind of weird.